misoc/integration/soc_core.py

```diff
@@ -135,9 +135,9 @@ def register_mem(self, name, origin, length, interface):
 
     def register_rom(self, interface, rom_size=0xa000):
         self.add_wb_slave(self.mem_map["rom"], rom_size, interface)
-        assert self.cpu_reset_address < rom_size
-        self.add_memory_region("rom", self.cpu_reset_address,
-                               rom_size-self.cpu_reset_address)
+        if not self.mem_map["rom"] <= self.cpu_reset_address < self.mem_map["rom"] + rom_size:
+            raise ValueError("CPU reset address 0x{:x} should be within ROM address range 0x{:x}-0x{:x}".format(self.cpu_reset_address,self.mem_map["rom"],self.mem_map["rom"] + rom_size ))
+        self.add_memory_region("rom", self.mem_map["rom"],rom_size)
```

Comment:
As explained, I don't think this is correct and the original code should be kept. IIRC the linker script relies on the "rom" memory region beginning at the reset address.

Reply:
Thanks you for the review I will consider other options. It looks like the requirement is that
mem_map["rom"] should be equal cpu_reset_address. Hence perhaps
assert self.cpu_reset_address == self.mem_map["rom"] is a good check(the result would remain the same).  The second "issue" on line 140 is that the size of the area is calculated as rom_size - self.cpu_reset_address and when the cpu_reset address is say 0x100 the size of the rom allocation becomes smaller.





self.add_memory_region("rom",self.cpu_reset_address,rom_size)

Reply:
> Hence perhaps
> assert self.cpu_reset_address == self.mem_map["rom"] is a good check

Sounds ok.

Reply:
> when the cpu_reset address is say 0x100 the size of the rom allocation becomes smaller.

Yes, that's intended, so that the linker does not overfill the ROM.

Reply:
If start address is say 0x7000 and rom size is 0x500 this value becomes negative.  I will do more experiments today to better understand what is going on. I am looking into this because I am trying to boot the rom from from spi (papilio pro) and the start address is not 0

Reply:
> If start address is say 0x7000 and rom size is 0x500 this value becomes negative.

Yes, the reset address should be within the ROM, obviously.

Reply:
But there's a bug in the original code too, it should be rom_size - (reset_address - rom_start).

Reply:
if reset_address is the same as the rom_start address the (reset_address - rom_start) no longer is needed